### DISCUSSION TITLE
How to retain Atom mappings in recursive SMARTS?

### DISCUSSION THREAD

**Discussion by MartinHoffmannJena** [rdkit/rdkit#8113] [Q&A] [answered]
Hi,

I have the following reaction SMARTs:

`rxn=rdChemReactions.ReactionFromSmarts("[$([#6:1]-[#8:2])&!$([#7]-[#6:1]-[#8:2])].[#8:3]-[#7:4]>>[#6:1]-[#8:2]-[#8:3]-[#7:4]")
`

The first part `[$([#6:1]-[#8:2])&!$([#7]-[#6:1]-[#8:2])]` describes an educt, which is a simple CO (in which the C has mapping index 1 and the O 2), but also imposes the condition, that no Nitrogen is allowed to be connected to that C atom (`&!$([#7]-[#6:1]-[#8:2])`).

The second part `[#8:3]-[#7:4]` then describes a simple ON bond, and the product side describes a COON bond between the two educts.

Now I run this reaction:

```
mol1 =Chem.MolFromSmiles("CCCCCCO")
mol4 =Chem.MolFromSmiles("ONNNNNN")

rxn=rdChemReactions.ReactionFromSmarts("[$([#6:1]-[#8:2])&!$([#7]-[#6:1]-[#8:2])].[#8:3]-[#7:4]>>[#6:1]-[#8:2]-[#8:3]-[#7:4]")

print(Chem.MolToSmiles(rxn.RunReactants((mol1,mol4))[0][0]))
```

and the output shows me 

```
COONNNNNN
[18:00:50] reactant 0 has no mapped atoms.
[18:00:50] product atom-mapping number 1 not found in reactants.
[18:00:50] product atom-mapping number 2 not found in reactants.
 ```

I would have expected the output to be `CCCCCCOONNNNNN` and the reason for that not being the case seems to be that my atom mappings in reactant 0 (`[$([#6:1]-[#8:2])&!$([#7]-[#6:1]-[#8:2])]`) got lost. I'm assuming this is the case because `[$(..)]` is treated as the sole reactant, which indeed doesnt have a mapping if you dont "expand" it back to the actual pattern it contains?

Is there any way to prevent this or a different way to formulate my SMARTS expression?

Thank you very much!

**Comment by bp-kelley** [comment, marked as answer]
Atom maps from the recursive level are not retained, they are only retained from the top level.

```
[$([#6:1]-[#8:2])&!$([#7]-[#6:1]-[#8:2])].[#8:3]-[#7:4]>>[#6:1]-[#8:2]-[#8:3]-[#7:4]
```

This isn't going to work as you expect, you can only have a mapping named once, and here you have 1 and 2 used twice at the recursive level, so they are ignored.

Think of recursive smarts as environments that need to be around the atom you are matching:

```
[$([#6]-[#8]):1]
```

matches a carbon attached to an oxygen.  Note that the atom map is outside the recursive smarts.

Hope this helps.

**Reply by MartinHoffmannJena** [reply]
Thank you very much, that did indeed help a lot.
